src/apps/vlan/vlan.lua

```diff
@@ -0,0 +1,174 @@
+module(..., package.seeall)
+
+local packet = require("core.packet")
+local bit = require("bit")
+local ffi = require("ffi")
+
+local C = ffi.C
+local receive, transmit = link.receive, link.transmit
+local cast = ffi.cast
+
+Tagger = {}
+Untagger = {}
+
+-- 802.1q
+local dotq_tpid = 0x8100
+local o_ethernet_ethertype = 12
+local uint32_ptr_t = ffi.typeof('uint32_t*')
+
+
+-- build a VLAN tag consisting of 2 bytes of TPID set to 0x8100 followed by the
+-- TCI field which in turns consists of PCP, DEI and VID (VLAN id). Both PCP
+-- and DEI is always 0
+local function build_tag(vid)
+   return ffi.C.htonl(bit.bor(bit.lshift(dotq_tpid, 16), vid))
```

Comment:
I know this code is moved from the lwAFTR but there still some bits that can be improved.

C is cached. In addition, `htonl`, `ntohl`, `htons` and `ntohs` functions are defined at `core/lib.lua`, so actually it's not necessary to access them via the C namespace. The `core/lib.lua` module is always included so it's not needed to require it.